### DISCUSSION TITLE
Update with filters

### DISCUSSION THREAD

**Discussion by louis030195** [supabase/supabase#14657] [Questions] [answered]
Hi

I'm trying to update some rows with `eq` filters (been using filters without issue many time before) and get a weird issue now.

I want to update a table `documents` where the rows have their `metadata.source` and `dataset_id` equal to something, I'm getting this error:

```
postgrest.exceptions.APIError: {'code': '42703', 'details': None, 'hint': None, 'message': 'column documents.source does not exist'}
```

But my query filters looks like this:

```
QueryParams('dataset_id=eq.unit_test&metadata.source=eq.notion.so')
```

So I don't know why it is talking about `documents.source`.

Here's a snippet of a part of my code:

```py
q = self.supabase.table("documents").update(
      [_d_update(row) for _, row in batch_df.iterrows()]
  )
# update only for this user id and dataset id if given
if user_id:
    q = q.eq("user_id", user_id)
if dataset_id:
    q = q.eq("dataset_id", dataset_id)
metadata_keys = list(where.keys())
metadata_values = list(where.values())
for key, value in zip(metadata_keys, metadata_values):
    q = q.eq(f"metadata.{key}", value)
return q.execute()
```

The whole code is open source [here](https://github.com/different-ai/embedbase/blob/dd531dd0af9db4894b434eb1148841851210956d/embedbase/database/supabase_db.py#L153=)

My table looks like this:

```sql
create table documents (
    id text primary key,
    data text,
    embedding vector (1536),
    hash text,
    dataset_id text,
    user_id text,
    metadata json,
    created_date TIMESTAMPTZ NOT NULL DEFAULT NOW()
);
```

[Ref](https://github.com/different-ai/embedbase/blob/feat/core/update-where/supabase/migrations/20230325092733_create_documents_table.sql)

What am I doing wrong?

Thanks a lot for your help 🙏

**Comment by louis030195** [comment]
Asked the question to GPT4 and the answer was:

Your code and table schema look correct, and the issue seems to be related to the `postgrest-py` package you are using to interact with your Supabase API. 

There seems to be a similar open issue on the `postgrest-py` GitHub repository which explains the exact problem you are encountering:
- [Issue: JSONB Column Filter Error](https://github.com/supabase/postgrest-py/issues/60)

Looking at the issue discussion, it appears you can use a workaround to avoid this error using the `.match()` function in combination with `eq`:

```python
for key, value in zip(metadata_keys, metadata_values):
    q = q.match({"metadata": {key: {"eq": value}}})
```

Replacing the `eq(f"metadata.{key}", value)` line with the above line should fix the error you are encountering. Give this a try and let me know if it works for your use case!

---

Which generated another issue:

```
postgrest.exceptions.APIError: {'code': '42883', 'details': None, 'hint': 'No operator matches the given name and argument types. You might need to add explicit type casts.', 'message': 'operator does not exist: json = unknown'}
```

(Gave up on gpt4 after worse answers following this)

**Comment by GaryAustin1** [comment]
Update with a foreign table filter does not appear to be supported currently.
https://github.com/PostgREST/postgrest/issues/2323
Which leaves an rpc call to a Postgres function or maybe a view (not sure on view).

**Reply by louis030195** [reply]
I don't understand "foreign table", here I'm filtering based on a property of the same table

But yes, I'll go with the SQL workaround if I can't use supabase sdk directly. An additional pain of creating a function though!

**Reply by GaryAustin1** [reply, marked as answer]
`metadata.source`  That is what made me believe you had a table called metadata with a column called source.   That is table.col like used for a foreign table filter.

I see now that is a json column....
You might try  `.eq('metadata->>source','value')`

**Reply by louis030195** [reply]
> `metadata.source` That is what made me believe you had a table called metadata with a column called source. That is table.col like used for a foreign table filter.
> 
> I see now that is a json column.... You might try `.eq('metadata->>source','value')`

ahhh, forgot this weird syntax, this was it

**Reply by steve-chavez** [reply]
The `->>` syntax comes from [PostgreSQL JSON operators](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-JSON-OP-TABLE). You can also use it with plain SQL(`select metadata->>'source'`).
